src/System.Sorting/src/TimSort.cs

```diff
@@ -0,0 +1,680 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Sorting;
+using System.Text;
+
+// This implementation was referenced from Java opensource, which has been found here:
+// http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8-b132/java/util/TimSort.java
+// Other Open Source references used C++:
+//https://github.com/gfx/cpp-TimSort/blob/master/timsort.hpp
+//Reference: https://en.wikipedia.org/wiki/Timsort
+//ref:http://svn.python.org/projects/python/trunk/Objects/listsort.txt - Original explanation from Tim
+//Timsort is a hybrid stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of partially sorted real-world data
+//Timsort is not a standalone algorithm but a hybrid, an efficient combination of a number of other algorithms
+//The algorithm finds subsets of the data that are already ordered, and uses that knowledge to sort the remainder more efficiently. This is done by merging an identified subset, called a run, with existing runs until certain criteria are fulfilled
+namespace System.Sorting
+{
+    //utility class for Timsort
+ internal static class TimSortUtilities
```

Comment:
I would just fold this stuff into TimSort itself.  Feels like better organization that way.